File: app/src/main/java/net/nashihara/naroureader/presenter/RankingRecyclerPresenter.java
package net.nashihara.naroureader.presenter;

import android.support.v4.util.Pair;
import android.text.TextUtils;
import android.util.Log;

import com.google.firebase.crash.FirebaseCrash;

import net.nashihara.naroureader.entities.NovelItem;
import net.nashihara.naroureader.views.RankingRecyclerView;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import narou4j.Narou;
import narou4j.Ranking;
import narou4j.entities.Novel;
import narou4j.entities.NovelRank;
import narou4j.enums.NovelGenre;
import narou4j.enums.OutputOrder;
import narou4j.enums.RankingType;
import rx.Emitter;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

public class RankingRecyclerPresenter implements Presenter<RankingRecyclerView> {

    private static final String TAG = RankingRecyclerPresenter.class.getSimpleName();

    private RankingRecyclerView view;

    public RankingRecyclerPresenter(RankingRecyclerView view) {
        attach(view);
    }

    @Override
    public void attach(RankingRecyclerView view) {
        this.view = view;
    }

    @Override
    public void detach() {
        view = null;
    }

    private void error(Throwable throwable) {
        view.showError();

        if (throwable == null) {
            return;
        }

        Log.e(TAG, "RankingRecyclerController: ", throwable.fillInStackTrace());
        FirebaseCrash.report(throwable);
    }

    public void fetchRanking(String rankingType) {
        if (TextUtils.isEmpty(rankingType)) {
            return;
        }

        if (rankingType.equals("all")) {
            fetchTotalRanking();
        } else {
            fetchEachRanking(RankingType.valueOf(rankingType));
        }
    }

    private void fetchTotalRanking() {
        fetchNovelsFromTotalRanking()
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(view::showRanking, this::error);
    }

    private Observable<List<NovelItem>> fetchNovelsFromTotalRanking() {
        return Observable.fromEmitter(emitter -> {
            Narou narou = new Narou();
            narou.setOrder(OutputOrder.TOTAL_POINT);
            narou.setLim(301);

            List<Novel> novels = null;
            try {
                novels = narou.getNovels();
            } catch (IOException e) {
                emitter.onError(e);
            }

            if (novels == null) {
                emitter.onError(null);
                emitter.onCompleted();
                return;
            }

            emitter.onNext(novelToNovelItem(novels));
            emitter.onCompleted();
        }, Emitter.BackpressureMode.NONE);
    }

    private List<NovelItem> novelToNovelItem(List<Novel> novels) {
        novels.remove(0);

        List<NovelItem> items = new ArrayList<>();
        NovelItem item;
        for (Novel novel : novels) {
            // ncodeが大文字と小文字が混在しているので小文字に統一
            novel.setNcode(novel.getNcode().toLowerCase());

            item = new NovelItem();
            NovelRank rank = new NovelRank();
            rank.setPt(novel.getGlobalPoint());
            item.setNovelDetail(novel);
            item.setRank(rank);
            items.add(item);
        }

        return items;
    }

    private void fetchEachRanking(RankingType rankingType) {
        fetchNovelRank(rankingType)
            .flatMap(map -> fetchPrevNovelRank(map, rankingType))
            .flatMap(this::setupNovelRanking)
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(view::showRanking, this::error);
    }

    private Calendar setupCalendarFromRankingType(RankingType rankingType) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());

        switch (rankingType) {
            case DAILY: {
                cal.add(Calendar.DAY_OF_MONTH, -2);
                break;
            }
            case WEEKLY: {
                cal.add(Calendar.DAY_OF_MONTH, -7);
                break;
            }
            case MONTHLY: {
                cal.add(Calendar.DAY_OF_MONTH, -31);
                break;
            }
            case QUARTET: {
                cal.add(Calendar.DAY_OF_MONTH, -31);
                break;
            }
        }

        return cal;
    }

    private Observable<HashMap<String, NovelItem>> fetchNovelRank(RankingType rankingType) {
        return Observable.fromEmitter(emitter -> {
            Ranking ranking = new Ranking();
            ArrayList<NovelRank> ranks = null;
            try {
                ranks = (ArrayList<NovelRank>) ranking.getRanking(rankingType);
            } catch (IOException e) {
                emitter.onError(e);
            }

            if (ranks == null) {
                emitter.onError(null);
                emitter.onCompleted();
                return;
            }

            emitter.onNext(novelRankToNovelMap(ranks, rankingType));
            emitter.onCompleted();
        }, Emitter.BackpressureMode.NONE);
    }

    private HashMap<String, NovelItem> novelRankToNovelMap(List<NovelRank> rankList, RankingType rankingType) {
        HashMap<String, NovelItem> map = new HashMap<>();
        for (NovelRank rank : rankList) {
            NovelItem novelItem = new NovelItem();
            rank.setRankingType(rankingType);
            novelItem.setRank(rank);
            map.put(rank.getNcode(), novelItem);
        }

        return map;
    }

    private Observable<HashMap<String, NovelItem>> fetchPrevNovelRank(HashMap<String, NovelItem> map, RankingType rankingType) {
        return Observable.fromEmitter(emitter -> {
            Ranking ranking = new Ranking();
            Calendar cal = setupCalendarFromRankingType(rankingType);

            List<NovelRank> rankList = null;
            try {
                rankList = ranking.getRanking(rankingType, cal.getTime());
            } catch (IOException e) {
                emitter.onError(e);
                return;
            }

            if (rankList == null) {
                emitter.onError(null);
                return;
            }

            setupRank(rankList, map);

            emitter.onNext(map);
            emitter.onCompleted();
        }, Emitter.BackpressureMode.NONE);
    }

    private void setupRank(List<NovelRank> rankList, HashMap<String, NovelItem> map) {
        for (NovelRank rank : rankList) {
            NovelItem item = map.get(rank.getNcode());

            if (item == null) {
                continue;
            }

            item.setPrevRank(rank);
            map.put(rank.getNcode(), item);
        }
    }

    private Observable<List<NovelItem>> setupNovelRanking(HashMap<String, NovelItem> map) {
        return Observable.fromEmitter(emitter -> {
            Pair<List<Novel>, Throwable> pair = fetchNovelFromNcode(map);

            if (pair.first == null) {
                error(pair.second);
                return;
            }

            List<NovelItem> items = validateNovels(pair.first, map);
            emitter.onNext(items);
            emitter.onCompleted();
        }, Emitter.BackpressureMode.NONE);
    }

    private Pair<List<Novel>, Throwable> fetchNovelFromNcode(HashMap<String, NovelItem> map) {
        Narou narou = new Narou();
        Set set = map.keySet();
        String[] array = new String[set.size()];
        set.toArray(array);

        narou.setNCode(array);
        narou.setLim(300);
        List<Novel> novels;
        try {
            novels = narou.getNovels();
        } catch (IOException e) {
            return Pair.create(null, e);
        }

        return Pair.create(novels, null);
    }

    private List<NovelItem> validateNovels(List<Novel> novelList, HashMap<String, NovelItem> map) {
        novelList.remove(0);

        ArrayList<NovelItem> items = new ArrayList<>();
        for (Novel novel : novelList) {
            NovelItem item = map.get(novel.getNcode());
            if (item != null) {
                // 大文字 → 小文字
                novel.setNcode(novel.getNcode().toLowerCase());

                item.setNovelDetail(novel);
                items.add(item);
            }
        }

        return items;
    }

    public void filterNovelRanking(List<NovelItem> novelItemList, NovelGenre[] filterIds,
                                   boolean[] itemChecked, String min, String max) {
        Set<NovelGenre> trueSet = new HashSet<>();
        List<NovelItem> filterList = new ArrayList<>();
        List<NovelItem> resultList = new ArrayList<>();

        int maxLength = validateCharLength(min);
        int minLength = validateCharLength(max);

        for (int i = 0; i < itemChecked.length; i++) {
            genreAndEndCheck(i, filterIds, itemChecked, novelItemList, trueSet, filterList);
        }

        for (int i = 0; i < filterList.size(); i++) {
            NovelItem target = filterList.get(i);

            if (!charLengthCheck(target, trueSet, maxLength, minLength)) {
                continue;
            }

            resultList.add(target);
        }

        view.showRanking(resultList);
    }

    private int validateCharLength(String length) {
        return length.equals("") ? 0 : Integer.parseInt(length);
    }

    private void genreAndEndCheck(int itemIndex, NovelGenre[] filterIds, boolean[] itemChecked,
                                  List<NovelItem> novelItemList, Set<NovelGenre> trueSet, List<NovelItem> filterList) {
        if (itemIndex == 0) {
            // 完結済チェック
            if (itemChecked[itemIndex]) {
                for (int j = 0; j < novelItemList.size(); j++) {
                    if (novelItemList.get(j).getNovelDetail().getIsNovelContinue() == 0) {
                        filterList.add(novelItemList.get(j));
                    }
                }
            }
            else {
                filterList.addAll(novelItemList);
            }

            return;
        }

        // ジャンルチェック
        if (itemChecked[itemIndex]) {
            trueSet.add(filterIds[itemIndex -1]);
        }
    }

    private boolean charLengthCheck(NovelItem target, Set<NovelGenre> trueSet, int max, int min) {
        if (!trueSet.contains(target.getNovelDetail().getGenre())) {
            return false;
        }

        // 文字数チェック
        if (max <= 0) {
            if (min > target.getNovelDetail().getNumberOfChar()) {
                return false;
            }
        } else {
            if (min > target.getNovelDetail().getNumberOfChar() || target.getNovelDetail().getNumberOfChar() > max) {
                return false;
            }
        }

        return true;
    }
}

File: app/src/main/java/net/nashihara/naroureader/presenter/RankingRecyclerPresenter.kt
package net.nashihara.naroureader.presenter

import android.util.Log

import com.google.firebase.crash.FirebaseCrash
import kotlinx.coroutines.experimental.Deferred
import kotlinx.coroutines.experimental.Job

import net.nashihara.naroureader.entities.NovelItem
import net.nashihara.naroureader.views.RankingRecyclerView

import java.util.ArrayList
import java.util.Calendar
import java.util.Date
import java.util.HashMap
import java.util.HashSet

import narou4j.Narou
import narou4j.Ranking
import narou4j.entities.Novel
import narou4j.entities.NovelRank
import narou4j.enums.NovelGenre
import narou4j.enums.OutputOrder
import narou4j.enums.RankingType
import net.nashihara.naroureader.addTo
import net.nashihara.naroureader.async
import net.nashihara.naroureader.ui

class RankingRecyclerPresenter(view: RankingRecyclerView) : Presenter<RankingRecyclerView> {

    private var view: RankingRecyclerView? = null

    private val jobList = mutableListOf<Job>()

    init {
        attach(view)
    }

    override fun attach(view: RankingRecyclerView) {
        this.view = view
    }

    override fun detach() {
        view = null
        jobList.forEach { it.cancel() }
    }

    private fun error(throwable: Throwable?) {
        view?.showError()

        if (throwable == null) {
            return
        }

        Log.e(TAG, "RankingRecyclerController: ", throwable.fillInStackTrace())
        FirebaseCrash.report(throwable)
    }

    fun fetchRanking(rankingType: String) {
        if (rankingType.isNullOrEmpty()) {
            return
        }

        if (rankingType == "all") {
            fetchTotalRanking()
        } else {
            fetchEachRanking(RankingType.valueOf(rankingType))
        }
    }

    private fun fetchTotalRanking() {
        ui {
            try {
                val novelList = fetchNovelsFromTotalRanking()
                val novelItemList = novelToNovelItem(novelList.await())
                view?.showRanking(novelItemList)
            } catch (e: Exception) {
                error(e)
            }
        }.addTo(jobList)
    }

    private fun fetchNovelsFromTotalRanking(): Deferred<MutableList<Novel>> {
        return async {
            val narou = Narou()
            narou.setOrder(OutputOrder.TOTAL_POINT)
            narou.setLim(301)
            return@async narou.novels
        }
    }

    private fun novelToNovelItem(novels: MutableList<Novel>): List<NovelItem> {
        novels.removeAt(0)

        val items = ArrayList<NovelItem>()
        var item: NovelItem
        for (novel in novels) {
            // ncodeが大文字と小文字が混在しているので小文字に統一
            novel.ncode = novel.ncode.toLowerCase()

            val rank = NovelRank()
            rank.pt = novel.globalPoint
            item = NovelItem(novelDetail = novel, rank = rank)
            items.add(item)
        }

        return items
    }

    private fun fetchEachRanking(rankingType: RankingType) {
        ui {
            try {
                val ranking = async { fetchNovelRank(rankingType) }
                val rankingWithNovelInfo = async { fetchRankingNovelItem(ranking.await()) }
                val setupedNovelRanking = async { setupPrevNovelRank(rankingWithNovelInfo.await(), rankingType) }.await()
                view?.showRanking(setupedNovelRanking)
            } catch (e: Exception) {
                error(e)
            }
        }.addTo(jobList)
    }

    private fun setupCalendarFromRankingType(rankingType: RankingType): Calendar {
        val cal = Calendar.getInstance()
        cal.time = Date()

        when (rankingType) {
            RankingType.DAILY -> {
                cal.add(Calendar.DAY_OF_MONTH, -2)
            }
            RankingType.WEEKLY -> {
                cal.add(Calendar.DAY_OF_MONTH, -7)
            }
            RankingType.MONTHLY -> {
                cal.add(Calendar.DAY_OF_MONTH, -31)
            }
            RankingType.QUARTET -> {
                cal.add(Calendar.DAY_OF_MONTH, -31)
            }
        }

        return cal
    }

    private fun fetchNovelRank(rankingType: RankingType): HashMap<String, NovelRank> {
        val ranks = Ranking().getRanking(rankingType)
        return novelRankToNovelMap(ranks, rankingType)
    }

    private fun novelRankToNovelMap(rankList: List<NovelRank>, rankingType: RankingType): HashMap<String, NovelRank> {
        val map = HashMap<String, NovelRank>()
        rankList.forEach {
            it.rankingType = rankingType
            map.put(it.ncode, it)
        }

        return map
    }

    private fun setupPrevNovelRank(novelItems: List<NovelItem>, rankingType: RankingType): List<NovelItem> {
        val cal = setupCalendarFromRankingType(rankingType)
        val rankList: List<NovelRank> = Ranking().getRanking(rankingType, cal.time)
        return setupPrevRank(rankList, novelItems)
    }

    private fun setupPrevRank(rankList: List<NovelRank>, novelItems: List<NovelItem>): List<NovelItem> {
        novelItems.map { novelItem ->
            novelItem.prevRank = rankList.find { rankItem ->
                rankItem.ncode.toLowerCase() == novelItem.novelDetail.ncode.toLowerCase() }
        }
        return novelItems
    }

    private fun fetchRankingNovelItem(map: HashMap<String, NovelRank>): List<NovelItem> {
        val narou = Narou()
        narou.setNCode(map.keys.toTypedArray())
        narou.setLim(300)

        val novels = narou.novels
        val novelItems = mutableListOf<NovelItem>()
        novels.map { novelItems.add(NovelItem(novelDetail = it, rank = map[it.ncode])) }

        return validateNovels(novels, map)
    }
    private fun validateNovels(novelList: List<Novel>, map: HashMap<String, NovelRank>): List<NovelItem> {
        val novelMutableList = novelList.toMutableList()
        novelMutableList.removeAt(0)

        val items = mutableListOf<NovelItem>()
        for (novel in novelMutableList) {
            val rank = map[novel.ncode]
            rank?.let {
                novel.ncode = novel.ncode.toLowerCase()
                items.add(NovelItem(novelDetail = novel, rank = it))
            }
        }

        return items
    }

    fun filterNovelRanking(novelItemList: List<NovelItem>, filterIds: Array<NovelGenre>,
                           itemChecked: BooleanArray, min: String, max: String) {
        val trueSet = HashSet<NovelGenre>()
        val filterList = ArrayList<NovelItem>()

        val maxLength = validateCharLength(min)
        val minLength = validateCharLength(max)

        for (i in itemChecked.indices) {
            genreAndEndCheck(i, filterIds, itemChecked, novelItemList, trueSet, filterList)
        }

        val resultList = filterList.indices
                .map { filterList[it] }
                .filter { charLengthCheck(it, trueSet, maxLength, minLength) }

        view?.showRanking(resultList)
    }

    private fun validateCharLength(length: String): Int {
        return when (length) {
            "" -> 0
            else -> Integer.parseInt(length)
        }
    }

    private fun genreAndEndCheck(itemIndex: Int, filterIds: Array<NovelGenre>, itemChecked: BooleanArray,
                                 novelItemList: List<NovelItem>, trueSet: MutableSet<NovelGenre>, filterList: MutableList<NovelItem>) {
        if (itemIndex == 0) {
            // 完結済チェック
            if (itemChecked[itemIndex]) {
                for (j in novelItemList.indices) {
                    if (novelItemList[j].novelDetail.isNovelContinue == 0) {
                        filterList.add(novelItemList[j])
                    }
                }
            } else {
                filterList.addAll(novelItemList)
            }

            return
        }

        // ジャンルチェック
        if (itemChecked[itemIndex]) {
            trueSet.add(filterIds[itemIndex - 1])
        }
    }

    private fun charLengthCheck(target: NovelItem, trueSet: Set<NovelGenre>, max: Int, min: Int): Boolean {
        if (!trueSet.contains(target.novelDetail.genre)) {
            return false
        }

        // 文字数チェック
        if (max <= 0) {
            if (min > target.novelDetail.numberOfChar) {
                return false
            }
        } else {
            if (min > target.novelDetail.numberOfChar || target.novelDetail.numberOfChar > max) {
                return false
            }
        }

        return true
    }

    companion object {

        private val TAG = RankingRecyclerPresenter::class.java.simpleName
    }
}
